datafusion/sql/tests/cases/plan_to_sql.rs

```diff
@@ -1406,3 +1414,150 @@ fn test_join_with_no_conditions() {
+        .parse_statement()?;
+    let state = MockSessionState::default();
+    let context = MockContextProvider { state };
+    let sql_to_rel = SqlToRel::new(&context);
+    let plan = sql_to_rel.sql_statement_to_plan(statement)?;
+
+    let udlp = MockUserDefinedLogicalPlan { input: plan };
+    let extension = LogicalPlan::Extension(Extension {
+        node: Arc::new(udlp),
+    });
+    let unparser =
+        Unparser::default().with_udlp_unparsers(vec![Arc::new(MockStatementUnparser {})]);
+    let sql = unparser.plan_to_sql(&extension)?;
+    let expected = "SELECT * FROM j1";
+    assert_eq!(sql.to_string(), expected);
+
+    if let Some(err) = plan_to_sql(&extension).err() {
+        assert_contains!(
+            err.to_string(),
+            "This feature is not implemented: Unsupported extension node: MockUserDefinedLogicalPlan");
+    } else {
+        panic!("Expected error");
+    }
+    Ok(())
+}
+
+struct MockSqlUnparser {}
+
+impl UserDefinedLogicalNodeUnparser for MockSqlUnparser {
+    fn unparse(
+        &self,
+        node: &dyn UserDefinedLogicalNode,
+        unparser: &Unparser,
+        _query: &mut Option<&mut QueryBuilder>,
+        _select: &mut Option<&mut SelectBuilder>,
+        relation: &mut Option<&mut RelationBuilder>,
+    ) -> Result<()> {
+        if let Some(plan) = node.as_any().downcast_ref::<MockUserDefinedLogicalPlan>() {
+            let Statement::Query(input) = unparser.plan_to_sql(&plan.input)? else {
+                return Ok(());
+            };
+            let mut derived_builder = DerivedRelationBuilder::default();
+            derived_builder.subquery(input);
+            derived_builder.lateral(false);
+            if let Some(rel) = relation {
+                rel.derived(derived_builder);
+            }
+        }
+        Ok(())
+    }
+}
+
+#[test]
+fn test_unparse_udlp_to_sql() -> Result<()> {
+    let dialect = GenericDialect {};
+    let statement = Parser::new(&dialect)
+        .try_with_sql("SELECT * FROM j1")?
+        .parse_statement()?;
+    let state = MockSessionState::default();
+    let context = MockContextProvider { state };
+    let sql_to_rel = SqlToRel::new(&context);
+    let plan = sql_to_rel.sql_statement_to_plan(statement)?;
+
+    let udlp = MockUserDefinedLogicalPlan { input: plan };
+    let extension = LogicalPlan::Extension(Extension {
+        node: Arc::new(udlp),
+    });
+
+    let plan = LogicalPlanBuilder::from(extension)
+        .project(vec![col("j1_id").alias("user_id")])?
+        .build()?;
+    let unparser =
+        Unparser::default().with_udlp_unparsers(vec![Arc::new(MockSqlUnparser {})]);
+    let sql = unparser.plan_to_sql(&plan)?;
+    let expected = "SELECT j1.j1_id AS user_id FROM (SELECT * FROM j1)";
+    assert_eq!(sql.to_string(), expected);
+
+    if let Some(err) = plan_to_sql(&plan).err() {
+        assert_eq!(
+            err.to_string(),
+            "External error: `relation` must be initialized"
```

Comment:
This error is expected?

Reply:
It's the original error for unsupported relation plan 🤔 I guess it could be improved by https://github.com/apache/datafusion/pull/13880#discussion_r1895366799